[1단계 방탈출 결제] 드라고(이용화) 미션 제출합니다.

Firedrago95 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  [방탈출 결제 프롤로그](https://prolog.techcourse.co.kr/studylogs/4426)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
[이번 미션에서 변경된 부분 🔗](https://github.com/woowacourse/spring-roomescape-payment/pull/177/files/1427301811ffb1b3ce3bda36c6238afe9a56905b..176b98f5d5b7d90d30fa8c38e61d67e72fe4da7e)

안녕하세요! 베루스 방탈출 결제 미션 리뷰 잘 부탁드립니다!
처음 외부 api를 호출하는 기능을 사용해보았습니다. `RestClient`를 사용하면서 외부 api를 호출할때 생각해보면 좋은 점이나 제가 놓친 부분들이 있는지 궁금합니다. 
 또한 실제 api를 호출하여 테스트하는 방법은 매번 결제 요청에 필요한 값을 가져와야하는 불편함이 있고, 운영 환경에 따라 테스트 결과가 달라질 수 있다는 단점이 있다고 판단하여 `MockRestServer`를 사용하여 api를 테스트 하였습니다. 베루스의 경우 외부 api를 테스트 해야하는 경우 어떤 기준으로 테스트 하는지 궁금합니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

Firedrago95 committed:
feat: migrate previous mission code

Firedrago95 committed:
docs: readme 추가

Firedrago95 committed:
feat: 클라이언트 코드 추가

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
docs: 결제기능 기능목록 작성

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
refactor: dto 필드 추가 및 유효성 검증 로직 추가

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

yeonnhuu committed:
feat: 예약 결제 기능 구현

Firedrago95 committed:
fix: 예약 결제 요청 로직 수정

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
feat: 예약 결제 요청 예외처리 로직 추가

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
refactor: 예약 결제 요청 예외처리 리팩토링

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

yeonnhuu committed:
refactor: 예약 결제 클라이언트 분리

Firedrago95 committed:
test: 결제 요청 테스트 작성

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
test: 결제요청 통합 테스트 수정

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
refactor: payment 패키지 구조 분리

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
Refactor: 예외 패키지 구조 변경

Firedrago95 committed:
feat: 결제 요청 타임아웃 기능 구현

verus-j commented on src/main/java/roomescape/reservation/presentation/dto/PaymentRequest.java:
```diff
@@ -0,0 +1,8 @@
+package roomescape.reservation.presentation.dto;
+
+public record PaymentRequest(
```
paymentKey, orderId, amount는 사용하지 않는 필드네요. 정의한 이유가 뭘까요??

Firedrago95 replied:
제가 생각할때 외부 결제 api를 호출하면서 필요한 세가지 정보를 담고 이를 전달하고 있는 역할을 하고 있는걸로 보이는데 혹시 `PaymentResponse`를 말씀하신 걸까요? `paymentResponse`의 경우 결제 성공시, 향후 결과를 db에 저장할 필요성이 있다고 생각하여 만들었습니다. `PaymentRequest`의 패키지 payment 안으로 이동하였습니다.

```java
@Override
    public PaymentResponse requestPayment(final PaymentRequest paymentRequest) {
        return restClient.post()
                .uri("/v1/payments/confirm")
                .contentType(MediaType.APPLICATION_JSON)
                .header("Authorization", "Basic " + encodeSecretKey())
                .body(paymentRequest)
                .retrieve()
                .onStatus(new RequestPaymentErrorHandler(new ObjectMapper()))
                .body(PaymentResponse.class);
    }
```
[refactor: PaymentRequest 클래스 위치 변경](https://github.com/woowacourse/spring-roomescape-payment/pull/177/commits/096e0e90593979afcb5aebd58e68b5f154d7891e)

verus-j replied:
아 바디로 전달하는군요! 저걸 제가 놓쳤네요

verus-j commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -70,6 +72,10 @@ public ReservationResponse createReservation(final ReservationRequest request, f
 
         Reservation reservation = Reservation.createWithoutId(request.date(), time, theme, member, Status.RESERVED);
         validateCanReserveDateTime(reservation, dateTime.now());
+
+        PaymentRequest paymentRequest = new PaymentRequest(request.paymentKey(), request.orderId(), request.amount());
+        paymentService.confirmPayment(paymentRequest);
+
```
외부 API 요청을 트랜잭션 내부에서 동작해도 괜찮을까요?

Firedrago95 replied:
외부 API가 트랜잭션 내부에서 함께 동작하게 되면 심각한 문제가 발생할 수 있다는 점을 확인했습니다.  
예를 들어, 결제 API가 먼저 호출되어 실제 결제가 이루어진 상황에서, 이후 DB 쓰기 작업 중 예외가 발생하면 결제는 되었지만 예약은 저장되지 않는 정합성 문제가 생길 수 있습니다. 이러한 문제를 방지하기 위해, DB 트랜잭션과 외부 API 호출을 분리했습니다.  
먼저 트랜잭션 내부에서 예약 정보를 저장하고, 커밋 이후에 결제 API를 호출하도록 변경했습니다. 다만 이 경우에도, 결제가 실패하거나 외부 API 호출이 실패할 수 있는 가능성을 고려하여  추후에는 결제 취소 요청 등 처리 방안을 도입할 수 있습니다.

[refactor: 예약 생성 로직 분리](https://github.com/woowacourse/spring-roomescape-payment/pull/177/commits/e3db2dbdee63b2acc2662074be0b6f6891b7ff65)

Firedrago95 replied:
`ReservationService` 내에서 외부 api 호출과 db저장을 함께 호출하여 내부호출로 인해 트랜잭션이 작동하지 않는 문제가 있었습니다. 스프링은 AOP를 통해 프록시 객체로 트랜잭션을 호출하게 되는데, 내부호출이 일어날 경우 프록시를 통한 호출이 적용되지 않아 트랜잭션이 적용되지 않습니다. 그래서 `ReservationWriterService`클래스에서 예약 저장 로직을 가지도록 분리하여 `ReservationService`에서 호출할때 트랜잭션이 적용될 수 있도록 수정 했습니다.

[refactor: 예약 생성 로직 분리 및 ReservationWriterService 도입](https://github.com/woowacourse/spring-roomescape-payment/pull/177/commits/7422140147aa33b8812423f367f7666eb993b9b0)

Firedrago95 replied:
`ReservationService`가 도메인 서비스를 의존하는 구조로 리팩토링 하였습니다.

[refactor: 예약 서비스 테스트 리팩토링 및 도메인 중심 설계 반영](https://github.com/woowacourse/spring-roomescape-payment/pull/177/commits/9fc02f0935a41f63fde0519ea7a15b6903399592)

verus-j replied:
드라고가 정리한 문제도 있지만, 트랜잭션 내부에서 외부 API 콜을 호출하는 경우 대기 시간동안 트랜잭션이 열려있는 문제가 생길 수 있습니다. 트랜잭션을 오래 열어두게 되면 커넥션을 오래 유지하기 때문에 다른 요청을 DB 커넥션을 가지지 못해 처리하지 못하는 문제가 생길 수도 있습니다. 이런 부분들도 고민해보면 좋을 것 같네요~

verus-j commented on src/test/java/roomescape/payment/infrastructure/TossPaymentClientTest.java:
```diff
@@ -0,0 +1,73 @@
+        .baseUrl("https://api.tosspayments.com")
+        .defaultStatusHandler(new RequestPaymentErrorHandler(new ObjectMapper()));
+
+    private MockRestServiceServer server = MockRestServiceServer.bindTo(testBuilder).build();
```
외부 API 요청을 MockRestServiceServer를 사용해 테스트 잘하셨네요

verus-j reviewed (changes requested):
안녕하세요 드라고
미션진행하시느라 고생하셨습니다.
피드백 남겨뒀으니 확인부탁드려요

Firedrago95 committed:
refactor: PaymentRequest 클래스 위치 변경

PaymentRequest를 적절한 패키지로 이동하여 모듈 간 의존성을 정리했습니다.

Firedrago95 committed:
refactor: 예약 생성 로직 분리

예약 생성 시 결제 확인 로직을 분리하여 가독성과 유지보수성을 개선

Firedrago95 committed:
refactor: 예약 생성 로직 분리 및 ReservationWriterService 도입

ReservationWriterService를 도입하여 예약 생성 관련 로직을 분리하고 Service의 코드 간소화를 구현.

Firedrago95 committed:
refactor: 예약 서비스 테스트 리팩토링 및 도메인 중심 설계 반영

기존 예약 서비스 테스트를 분리하고 도메인 중심 설계를 반영하여 WaitingDomainService와 ReservationDomainService를 추가.

Firedrago95 committed:
refactor: 불필요한 필드 제거 및 코드 개선

ReservationService 내 중복 필드 제거 및 가독성 개선을 위한 리팩토링.

verus-j commented on src/test/java/roomescape/reservation/service/WaitingDomainServiceTest.java:
```diff
@@ -0,0 +1,184 @@
+        assertThat(secondResponse.time().id()).isEqualTo(1L);
+    }
+
+    static class WaitingDomainConfig {
```
별도로 config 구성한 이유가 있을까요?

verus-j reviewed (approved):
안녕하세요 드라고
피드백 반영해주신 것 확인했습니다.
이번 단계는 여기서 이만 머지하겠습니다~
